libroach: add missing destructor (in parent class)

knz opened the pull request:
I think this is slightly better future-proof (you can carry `unique_ptr<KeyManager>` with this)

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21497)
<!-- Reviewable:end -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

This is good C++ style. As a reminder for others, any C++ class which declares virtual functions should declare a virtual destructor (or inherit from a class which declares a virtual destructor). I don't recall all of the badness that results from not doing so, but recall that it is there. Something about the destructor not being called properly. 

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21497#-:-L30alv-Yu67A2BvKesZ:b-dn23zg)*
<!-- Sent from Reviewable.io -->

knz committed:
libroach: add missing destructor (in parent class)

mberhault commented on c-deps/libroach/ccl/key_manager.h:
```diff
@@ -85,8 +87,6 @@ class FileKeyManager : public KeyManager {
                           const std::string& old_key_path)
       : env_(env), active_key_path_(active_key_path), old_key_path_(old_key_path) {}
 
-  virtual ~FileKeyManager();
```
That should stay for any direct users of `FileKeyManager`

knz replied:
No it doesn't: the parent class of `FileKeyManager` already has a virtual destructor (with my change above), so automatically all derived classes have their destructor virtual too.

mberhault replied:
Sure, but it's still weird to have `~FileKeyManager()` in the `.cc` that isn't reflected in the `.h`.

knz replied:
oh I see, sorry for that. I am so used to template programming with all code in headers I hadn't even looked at the cc file. 

Fixed, PTAL.

mberhault commented:
If the abstract class does not have a virtual destructor, then anyone holding a pointer to the abstract class will not call the derived class' destructor (I'm not doing that yet, but I will soon). We still want the virtual destructor on the derived class though.

mberhault commented:
LGTM. Thanks for the fix. Small nit: you can leave the actual implementation (`~KeyManager(){}`) in the `.h`, the reason to avoid inline destructors is when they may be too big, but `KeyManager` doesn't have anything to cleanup.

knz commented:
> Small nit: you can leave the actual implementation (~KeyManager(){}) in the .h, the reason to avoid inline destructors is when they may be too big

The same argument would apply to `~FileKeyManager()`, doesn't it? If I do it for one then the other should be the same for consistency.

In general if the class is not a template i'd rather have the definition in a cc file to avoid linker churn. The churn is the same regardless of function body size.

knz commented:
The argument about inlining just makes me think...

I now realize there is a bunch of `virtual` methods on *non-template* classes in the header file. AFAIK virtual methods cannot be inlined in most cases. I do not see that there is a benefit to have their definition in the header file (again, non-template). In fact I think it adds linker churn without any upside. What do you think?

knz commented:
I appended 2 commits to the PR to clarify what I mean. PTAL.

mberhault commented on c-deps/libroach/ccl/key_manager.h:
```diff
@@ -74,6 +74,8 @@ class KeyManager {
   // **WARNING**: this must not be logged, displayed, or stored outside the key registry.
   virtual std::unique_ptr<enginepbccl::SecretKey> GetKey(const std::string& id) = 0;
+
+  virtual ~KeyManager();
```
tiny style nit: the destructor declaration should be at the beginning of the `public:` section.

knz replied:
Done

mberhault commented:
LGTM. Fair enough.

knz committed:
libroach: make non-overloaded methods non-virtual

... so that they can be reliably inlined.

Release note: none

knz committed:
libroach: define the body of virtual methods in the cc file

... since virtual methods cannot be readily inlined; to avoid linker churn.

Release note: none
